pkg/controller/operators/adoption_controller.go

```diff
@@ -0,0 +1,409 @@
+		Watches(&source.Kind{Type: &corev1.ServiceAccount{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &corev1.ConfigMap{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.Role{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.RoleBinding{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.ClusterRole{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &rbacv1.ClusterRoleBinding{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &apiextensionsv1.CustomResourceDefinition{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &apiregistrationv1.APIService{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &operatorsv1alpha1.Subscription{}}, enqueueCSV).
+		Watches(&source.Kind{Type: &operatorsv1alpha1.InstallPlan{}}, enqueueCSV).
+		Complete(reconcile.Func(r.ReconcileClusterServiceVersion))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewAdoptionReconciler constructs and returns an AdoptionReconciler.
+// As a side effect, the given scheme has operator discovery types added to it.
+func NewAdoptionReconciler(cli client.Client, log logr.Logger, scheme *runtime.Scheme) (*AdoptionReconciler, error) {
+	// Add watched types to scheme.
+	if err := AddToScheme(scheme); err != nil {
+		return nil, err
+	}
+
+	factory, err := decorators.NewSchemedOperatorFactory(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AdoptionReconciler{
+		Client: cli,
+
+		log:     log,
+		factory: factory,
+	}, nil
+}
+
+// ReconcileSubscription labels the CSVs installed by a Subscription as components of an operator named after the subscribed package and install namespace.
+func (r *AdoptionReconciler) ReconcileSubscription(req ctrl.Request) (reconcile.Result, error) {
+	// Set up a convenient log object so we don't have to type request over and over again
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling subscription")
+
+	// Fetch the Subscription from the cache
+	ctx := context.TODO()
+	in := &operatorsv1alpha1.Subscription{}
+	if err := r.Get(ctx, req.NamespacedName, in); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Could not find Subscription")
+		} else {
+			log.Error(err, "Error finding Subscription")
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// OLM generated Operators are named after their packages and further qualified by the install namespace
+	if in.Spec == nil || in.Spec.Package == "" {
+		log.Info("subscription missing status, ignoring")
+		return reconcile.Result{}, nil
+	}
+
+	// Wrap with convenience decorator
+	operator, err := r.factory.NewPackageOperator(in.Spec.Package, in.GetNamespace())
+	if err != nil {
+		log.Error(err, "Could not wrap Operator with convenience decorator")
+		return reconcile.Result{}, nil
+	}
+
+	// Find the Subscription's CSVs and apply the component label if necessary
+	csv := &operatorsv1alpha1.ClusterServiceVersion{}
+	if name := in.Status.InstalledCSV; name != "" {
```

Comment:
Just double-checking, does `InstalledCSV` get populated if the Subscription fails to install the CSV? I think we want to surface the Operator object regardless of whether the Subscription succeeds or not right? Just want to make sure we aren't limiting this to only successful CSV installations -- if the install fails the Operator still surfaces useful information to the end user.